src/renderer/features/network/CustomRpcForm/model/add-custom-rpc-model.ts

```diff
@@ -0,0 +1,179 @@
+import { createEffect, createEvent, createStore, sample } from 'effector';
+import { createForm } from 'effector-forms';
+
+import { networkService, RpcValidation } from '@shared/api/network';
+import { ExtendedChain } from '@entities/network';
+import { CustomRpcForm, NodeExistParam, RpcCheckResult } from '../lib/types';
+import { manageNetworkModel } from '@pages/Settings/Networks/model/manage-network-model';
+import { fieldRules } from '../lib/utils';
+
+const $addCustomRpcForm = createForm({
+  fields: {
+    name: {
+      init: '',
+      rules: fieldRules.name,
+    },
+    url: {
+      init: '',
+      rules: fieldRules.url,
+    },
+  },
+  validateOn: ['blur'],
+});
+
+const formInitiated = createEvent();
+const networkChanged = createEvent<ExtendedChain>();
+const rpcConnectivityChecked = createEvent<RpcCheckResult>();
+const nodeExistChecked = createEvent<boolean>();
+const processStarted = createEvent<boolean>();
+
+const $selectedNetwork = createStore<ExtendedChain | null>(null);
+const $rpcConnectivityResult = createStore<RpcCheckResult>(RpcCheckResult.INIT);
+const $isNodeExist = createStore<boolean>(false);
+const $isProcessStarted = createStore<boolean>(false);
+
+const checkRpcNodeFx = createEffect(
+  async ({ network, url }: { network: ExtendedChain | null; url: string }): Promise<RpcCheckResult | null> => {
+    if (!network) return null;
```

Comment:
This is covered already, but the typing inferred isn't aware of it, I personaly prefer to have these 2 ifs, rather than a `!`. Also if I change the types of the effect, then I get another TS issue that the effect should extend the types of the source. Let me know what outcome you'd prefer.

Reply:
You can use `network: ExtendedChain` type without null, and filter it sample. 

Example
```typescript
sample({
    clock: getConfigFx.doneData,
    filter: (config: XcmConfig | null): config is XcmConfig => Boolean(config),
    target: calculateFinalConfigFx,
});
```

Reply:
`sample` is doing one thing and do it really well, so it's the best place to place filtering (with `filter`), data mapping (with `fn`).
And if we managed to reach the following `effect` (in `target`) it must receive all required data to proceed and not fast return on the first line, otherwise it's a red flag and better to step back and design more robust data flow.

Reply:
Thanks for the pointers, after fighting a bit more, I ended up having TS happy. I'd say it's pretty hard to read though :sweat: 

```ts
sample({
  clock: $addCustomRpcForm.submit,
  source: { network: $selectedNetwork, url: $addCustomRpcForm.fields.url.$value },
  filter: (params: { network: ExtendedChain | null; url: string }): params is CheckRpcNodeFxParams => !!params.network,
  target: [checkRpcNodeFx, isNodeExistFx],
});
```